### PR TITLE
[WIP] Add v6 cmd example spike

### PR DESCRIPTION
**Pull request by RasPhilCo**
```
$ heroku kafka:topics:create parrothead1 --app margaritaville --retention-time 10days
```

### CODE REVIEW

**Comment by uhoh-itsmaciek** [comment]
?

File: commands/topics_create_v6.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
+'use strict';
+
+const Command = require('cli-engine-command').default
+const AppFlag = require('cli-engine-command/lib/flags/app').AppFlag
+const BooleanFlag = require('cli-engine-command/lib/flags/boolean').default
+const StringFlag = require('cli-engine-command/lib/flags/string').default
```
Comment (uhoh-itsmaciek) [inline commands/topics_create_v6.js]: I really like this approach, but this seems really verbose... Can we do something like

```javascript
const flags = require('cli-engine-command/lib/flags')
// ...and then...
TopicCreate.flags = {
  app: flags.app({ required: true}),
  partitions: flags.string({ description: 'number of partitions to give the topic' }),
  'replication-factor': flag.string({ description: 'number of replicas the topic should be created across' }),
  'retention-time': flag.duration(),
  compaction: flag.boolean({ description: 'whether to use compaction for this topic' })
}
```

?

Comment (RasPhilCo) [inline commands/topics_create_v6.js]: That already exists. I was being verbose to demonstrate. Will update.

Comment (dickeyxxx) [inline commands/topics_create_v6.js]: the `.default` thing is a pain point because we're moving to using the import syntax with babel, but in existing plugins that won't be using the import syntax and this is how babel exports it.

But we do actually export a flag index file from lib/flags: https://github.com/heroku/cli-engine-command/blob/master/src/flags/index.js

though it isn't used much, and I think they should be renamed to match what you have above. (They use the CamelCase because in an earlier iteration these were classes before I decided they were better off as POJOs).

Reason I haven't used that much is because I thought it would look a little weird to have a mix of flags from different places. DurationFlag wouldn't be a part of cli-engine-command, so it would look like this if we used the index:

```js
const flags = require('cli-engine-command/lib/flags')
const DurationFlag = require('../lib/flags/duration')

// ...and then...
TopicCreate.flags = {
  app: flags.app({ required: true}),
  partitions: flags.string({ description: 'number of partitions to give the topic' }),
  'replication-factor': flag.string({ description: 'number of replicas the topic should be created across' }),
  'retention-time': DurationFlag(),
  compaction: flag.boolean({ description: 'whether to use compaction for this topic' })
}
```

which I thought looked a little strange as opposed to:

```js
// ignoring the ugly imports

// ...and then...
TopicCreate.flags = {
  app: StringFlag({ required: true}),
  partitions: StringFlag({ description: 'number of partitions to give the topic' }),
  'replication-factor': StringFlag({ description: 'number of replicas the topic should be created across' }),
  'retention-time': DurationFlag(),
  compaction: BooleanFlag({ description: 'whether to use compaction for this topic' })
}
```

Comment (uhoh-itsmaciek) [inline commands/topics_create_v6.js]: Good point. Maybe we should do

```javascript
const flags = Object.assign({ duration: require('../lib/flags/duration') }, require('cli-engine-command/lib/flags'))
```
...

Ok, I'll show myself out.

```

Comment (uhoh-itsmaciek) [inline commands/topics_create_v6.js]: But more seriously, re: naming--I'm not a big Go fan, but I really like its [naming guidelines](https://golang.org/doc/effective_go.html#package-names)--specifically, that you are encouraged to scope the semantic names of the functions in your package to take advantage of the package qualifier (e.g., `bufio.Reader` versus `bufio.BufReader` as in their examples). I think it results in concise, meaningful names. Not a big deal, but thought I'd bring it out since that's the only thing that jumped out at me. I agree the lack of possibility of consistency with custom flags is unfortunate, but then custom flags should be pretty rare...

Comment (dickeyxxx) [inline commands/topics_create_v6.js]: jumping off that... in `../lib/flags/duration.js`:

```js
const base = require('cli-engine-command/lib/flags')
const duration = function () {/* etc */}
module.exports = Object.assign({duration}, base)
```

Comment (uhoh-itsmaciek) [inline commands/topics_create_v6.js]: Heh, okay, that kinda works. Maybe that wasn't entirely crazy.

Comment (dickeyxxx) [inline commands/topics_create_v6.js]: and yeah, I much prefer the `group.type` naming versus putting it in the name of the thing

**Comment by RasPhilCo** [comment]
- [ ] When https://github.com/heroku/cli-engine-command/pull/25 merged, update `cli-engine-command` dep

File: commands/topics_create_v6.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+const flags = require('../lib/flags')
+
+class TopicCreate extends Command {
+  async run() {
+    this.out.log(`running create with retention-time: ${this.flags['retention-time']} ms`)
+  }
+}
+
+exports.default = TopicCreate
+TopicCreate.topic = 'kafka'
```
Comment (dickeyxxx) [inline commands/topics_create_v6.js]: shitty we can't use the class properties here

File: commands/topics_create_v6.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+const Command = require('cli-engine-command').default
+const flags = require('../lib/flags')
+
+class TopicCreate extends Command {
+  async run() {
+    this.out.log(`running create with retention-time: ${this.flags['retention-time']} ms`)
+  }
+}
+
+exports.default = TopicCreate
```
Comment (dickeyxxx) [inline commands/topics_create_v6.js]: you could do `exports.cmd = TopicCreate` to better match the existing commands

File: lib/flags/duration.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+'use strict';
+
+const parseDuration = require('../shared').parseDuration
+
+module.exports = function DurationFlag (options = {}, env = process.env) {
```
Comment (dickeyxxx) [inline lib/flags/duration.js]: you don't need to have `env = process.env` for this one

File: package.json [hunk 5]
```diff
@@ -21,7 +22,8 @@
     "humanize-plus": "^1.8.2",
     "lodash.uniqby": "4.6.1",
     "no-kafka": "github:hunterloftis/kafka",
-    "node-spinner": "0.0.4"
+    "node-spinner": "0.0.4",
+    "snappy": "6.0.0"
```
Comment (dickeyxxx) [inline package.json]: why is this added? isn't this a dependency of `no-kafka`?

File: index.js [hunk 6]
```diff
@@ -23,6 +23,7 @@ exports.commands = [
   // new versions of commands with deprecated variants
   require('./commands/topics.js').cmd,
   require('./commands/topics_create.js').cmd,
```
Comment (dickeyxxx) [inline index.js]: I think you should actually replace the topics_create implementation for this PR

File: commands/topics_create_v6.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  `;
+TopicCreate.args = [
+  {name: 'TOPIC'},
+  {name: 'CLUSTER', optional: true}
+]
+TopicCreate.flags = {
+  app: flags.app({required: true}),
+  partitions: flags.string({description: 'number of partitions to give the topic'}),
+  'replication-factor': flags.string({description: 'number of replicas the topic should be created across'}),
+  'retention-time': flags.duration(),
```
Comment (dickeyxxx) [inline commands/topics_create_v6.js]: did this not have a description before?

Comment (uhoh-itsmaciek) [inline commands/topics_create_v6.js]: The description is embedded in the custom flag handling. It's probably a good idea to pull it out and pass it in here instead.

Comment (dickeyxxx) [inline commands/topics_create_v6.js]: ah, because it says "in the topic" in the default description I agree. I think for some custom flags though it would be nice to give the description a default value

**Comment by uhoh-itsmaciek** [comment]
So the idea here is to give an example implementation of a command using the v6 APIs for us, yes? Is making the changes necessary to be able to run in v6, or is this just a nice-to-have?

**Comment by dickeyxxx** [comment]
v5 syntax will work fine in v6, no need to convert anything unless you want new features like the flags. This PR is sort of a proving ground for us to validate what we've come up with. I think it's in a pretty good place to give us the shareable flag concept I've been looking for.

**Comment by dickeyxxx** [comment]
Notably you could just provide a parse function to a string flag and it would do basically the same thing (like your original proposal)

**Comment by dickeyxxx** [comment]
Would definitely like to know if you have any thoughts on the general concept @uhoh-itsmaciek

**Comment by uhoh-itsmaciek** [comment]
Ok, awesome. Let me sleep on it and I'll give some feedback tomorrow.

File: commands/topics_create_v6.js (hunk 2 above)
Comment (uhoh-itsmaciek) [inline commands/topics_create_v6.js]: Like

```javascript
class TopicCreate extends Command {
  static topic = 'kafka'
}
```
? How come? Is that not a thing yet (I'm not that familiar with what's standard these days; just what we're using in DHC)?

Can we do something like

```javascript
Object.assign(TopicCreate, {
  topic: 'kafka',
  command: 'topics:create',
  ...
})
```
instead? That's not that much less typing but feels a little better organized than a series of assignments.

File: commands/topics_create_v6.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+  $ heroku kafka:topics:create page-visits --partitions 100
+  $ heroku kafka:topics:create page-visits kafka-shiny-2345 --partitions 100 --replication-factor 3 --retention-time 10d
+  $ heroku kafka:topics:create page-visits kafka-shiny-2345 --partitions 100 --compaction
+  `;
+TopicCreate.args = [
+  {name: 'TOPIC'},
+  {name: 'CLUSTER', optional: true}
+]
+TopicCreate.flags = {
+  app: flags.app({required: true}),
```
Comment (uhoh-itsmaciek) [inline commands/topics_create_v6.js]: Is `flags.app` only intended for the "main" `--app` flag, or can I use it multiple times if my command operates on multiple apps?

File: commands/topics_create_v6.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+  partitions: flags.string({description: 'number of partitions to give the topic'}),
+  'replication-factor': flags.string({description: 'number of replicas the topic should be created across'}),
```
Comment (uhoh-itsmaciek) [inline commands/topics_create_v6.js]: These could be `flags.number` or something, right? The current command doesn't validate they're numbers because Shogun will anyway, but if it's easy to do, might as well handle this in the client.

File: commands/topics_create_v6.js (hunk 7 above)
Comment (uhoh-itsmaciek) [inline commands/topics_create_v6.js]: Yeah, that makes sense.

Comment (uhoh-itsmaciek) [inline commands/topics_create_v6.js]: I am super-excited about "typed" args and custom type support, by the way :1st_place_medal:.

File: package.json (hunk 5 above)
Comment (uhoh-itsmaciek) [inline package.json]: It is; I'm curious as well.

File: lib/flags/duration.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+  const defaultOptions = {
+    description: 'length of time messages in the topic should be retained (at least 24h)',
+    parse: (input) => {
+      let retentionTimeMillis = parseDuration(input)
+      if (!retentionTimeMillis) {
+        throw new Error(`Could not parse retention time '${input}'; expected value like '10d' or '36h'`)
```
Comment (uhoh-itsmaciek) [inline lib/flags/duration.js]: How does error handling for flag parsing work? It occurs to me that the way I'm doing it currently, either this flag type is tightly coupled to retention time (as per the error message above) or the error message will be generic ("`Could not parse duration ${input}...`"). I think the CLI framework itself could potentially handle this by catching errors from the flag parser and re-throwing with more information. It knows the flag name and the input, so you could easily produce errors like `Could not parse ${flag_name} ${input}: ${parseError.message}` or something. This avoids shoving too much info into the parser just to have it generate messages, but still gives clear and informative errors. Thoughts?

**Comment by dickeyxxx** [comment]
for some reason I can't comment on your class properties comment, but you do have the syntax we use in v6: https://github.com/heroku/cli-engine/blob/master/src/commands/help.js#L12

but it requires babel for now. Until the class properties land I like your Object.assign method a lot more than this. I'll update this PR to use that so we have a nice example.

File: lib/flags/duration.js (hunk 10 above)
Comment (dickeyxxx) [inline lib/flags/duration.js]: great idea. I think the syntax within the flag can stay the same probably here (just throwing an exception on an error), but the parser could trap it and add the extra information.

Comment (dickeyxxx) [inline lib/flags/duration.js]: created https://github.com/heroku/cli-engine/issues/127 to track

File: commands/topics_create_v6.js (hunk 9 above)
Comment (dickeyxxx) [inline commands/topics_create_v6.js]: yeah definitely, and then you'd get the typing if the plugin was using flow

File: commands/topics_create_v6.js (hunk 8 above)
Comment (dickeyxxx) [inline commands/topics_create_v6.js]: while this would technically work:

```js
flags = {
  'source-app': flags.app({required: true}),
  'destination-app': flags.app()
}
```

it would behave strangely since this flag defaults to the git remote. Maybe that's what you would want, but I think generally you would probably want to do:

```js
flags = {
  'source-app': flags.app({required: true}),
  'destination-app': flags.string()
}
```

so that the defaulting to app in git remotes only happens on one flag.

File: .travis.yml [hunk 11]
```diff
@@ -1,6 +1,6 @@
 language: node_js
 node_js:
-  - "6.2.2"
+  - "7.7.1"
```
Comment (uhoh-itsmaciek) [inline .travis.yml]: Awwww yisssss

File: commands/topics_create.js [hunk 12]
```diff
@@ -1,92 +1,87 @@
...
     }
-  }
-  let compaction = flags['compaction'] || false
+    const heroku = this.heroku
 
-  yield withCluster(heroku, context.app, context.args.CLUSTER, function * (addon) {
-    let addonInfo = yield fetchProvisionedInfo(heroku, addon)
+    let flags = context.flags
+    let retentionTimeMillis = flags['retention-time']
+    let compaction = flags['compaction'] || false
```
Comment (uhoh-itsmaciek) [inline commands/topics_create.js]: That reminds me, no way to specify defaults in the flag definition?

File: lib/command.js [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+    let options = {
+      userAgent: h.requestOptions.headers['user-agent'],
+      debug: process.env.HEROKU_DEBUG,
+      token: h.auth,
+      host: `${h.requestOptions.protocol}//${h.requestOptions.host}`
+    }
+    this._heroku = new HerokuClient(options)
+    return this._heroku
+  }
+}
```
Comment (uhoh-itsmaciek) [inline lib/command.js]: I'm surprised by this. Do we need this boilerplate in every plugin?

File: commands/topics_create.js [hunk 14]
```diff
@@ -1,92 +1,87 @@
...
-    retentionTimeMillis = parseDuration(flags['retention-time'])
-    if (!retentionTimeMillis) {
-      cli.exit(1, `Could not parse retention time '${flags['retention-time']}'; expected value like '10d' or '36h'`)
+class TopicsCreate extends Command {
+  async run () {
+    const context = {
+      app: this.flags.app,
+      args: {TOPIC: this.argv[0], CLUSTER: this.argv[1]},
+      flags: this.flags
     }
```
Comment (uhoh-itsmaciek) [inline commands/topics_create.js]: Why does the command do this rather than being passed this info like before?

File: lib/command.js (hunk 13 above)
Comment (dickeyxxx) [inline lib/command.js]: something we uncovered in this exercise we want to fix. We're moving away from the current node client to the more general `http-call`. It's not yet fully compatible though so until then we need something like this to get the old client.

We'll have a standard base heroku command class that we can abstract it down to though, so the plugin itself won't have to define it.

Comment (uhoh-itsmaciek) [inline lib/command.js]: Ok, nice.

File: commands/topics_create.js (hunk 14 above)
Comment (dickeyxxx) [inline commands/topics_create.js]: Not sure if you're asking about why we're using this or why these changed format a bit, but I can explain both:

Something I've gone back and forth on (whether or not run should take any arguments) for a bit I considered having the args and flags be arguments run could expect but ultimately it seemed best to include everything on `this`. I don't have a super strong preference, but part of it was that it was simpler to write v6 commands with flow since lots of the properties are already typed. It also makes it more extensible since changing the run arguments would be hard with flow, but it's easy to add new properties to a class (like if another CLI wanted a totally different flag parser or something we could support that by having them place their properties somewhere else like `this.myparser.flags`).

As far as these specific properties, `context.app` was not ported to `this.app` because I wanted to remove all heroku-specific things from the engine.

`context.args` was a little funny before. If variableArgs was set it was an array, and if it was false it was this map. The map was useful, but difficult to get to work with flow typing. I decided to make that more consistent by just always returning an array. I considered having both, one in `this.argv` and another in `this.args` where one was an object and the other an array, but that just seemed like I was making it more confusing.

I don't think most code will create a context object like this. It was just done to make an object that had everything the existing code expected.

We originally thought we would just leave all the v5 code as-is and not worry about porting things over. But some of the functionality like the new flags are useful enough to use, but this code to make a v5 command work in v6 is annoying. Hopefully we can make the v6 heroku base class (which we haven't written yet) compatible with a lot of this stuff. Maybe give it the ability to create a v5-compatible context object and a heroku-client instance we could pass to existing code.

File: commands/topics_create.js (hunk 12 above)
Comment (dickeyxxx) [inline commands/topics_create.js]: we have defaults now!

File: commands/topics_create.js (hunk 14 above)
Comment (uhoh-itsmaciek) [inline commands/topics_create.js]: Ok, I think that makes sense. My main complaint was that this seems like a lot of boilerplate for the command to handle (well, okay, just the args--referencing the other stuff directly from this.flags seems fine). Given that the vast majority of commands use fixed args, it would be nice to have a mechanism that doesn't make you repeat the arg specification from the command definition but still gives you a nice shorthand access to named arguments rather than digging through argv. You can probably do something like

```javascript
this.args = this.constructor.args
  .map((arg, i) => [ arg, i ])
  .reduce((accum, [ arg, i]) => { accum[arg.name] = this.argv[i] })
```

behind the scenes before invoking `run` in those cases (although uh no idea how that works with flow if at all).

Comment (dickeyxxx) [inline commands/topics_create.js]: glad you mentioned, and I think you're right. My first thought was you could easily put this at the top of your run function:

```js
let [foo, bar] = this.args
```

But you're right, you've already defined the args above, you should be able to reference them once even if they're fixed. I was thinking maybe we could go back to the whole object or map thing based on whether or not variableArgs is set, but making that work in a typesafe way seems really hard if even possible. And I'd like it to remain typesafe so you're sure everything in the command is configured right.

your example would work great, but we need support for grabbing the args as an array as well.

at one point I had this:

```js
class Cmd {
  static argv: sting[]
  static args: {[name: string]: string}
}
```

so you could access them either way. This allows you to also have named args as well as variable args in the same command. (something akin to `function (firstarg, ...otherArgs)`)

cc @heroku/cli

Comment (uhoh-itsmaciek) [inline commands/topics_create.js]: Good call regarding array destructuring, but I think it's still a bit too much work. I like the `argv` plus `args` approach.

Comment (dickeyxxx) [inline commands/topics_create.js]: https://github.com/heroku/cli-engine-command/pull/32

**Comment by dickeyxxx** [comment]
reminder to
- [ ] update with cli-engine-heroku updates

**Comment by idan** [comment]
Hey @dickeyxxx  — what's the next step here?

**Comment by dickeyxxx** [comment]
next steps are to update with the new base class I've created, then likely a long-ish wait until v6 is fully released at which point we will be at a place we can merge it

**Comment by dickeyxxx** [comment]
ok base class added, but it depends on fixing this minor issue: https://github.com/heroku/cli-engine-heroku/pull/8

**Comment by idan** [comment]
@dickeyxxx looks like that got merged — can this move forward?

**Comment by dickeyxxx** [comment]
yes, I won't be able to work on it this week, but hopefully soon. There is a chance @RasPhilCo may get to look at it first

**Comment by RasPhilCo** [comment]
This is too out-of-date to save. Suggest closing and revisiting converting to v6 at a later date.

**Comment by jdxcode** [comment]
yeah the point of this was more the exercise than the code so let's just drop it
